### PR TITLE
[Cache] Count cache hits/misses in ProxyAdapter

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #17537 partially |
| License | MIT |
| Doc PR | - |

I propose to add this subset of the `Doctrine\Common\Cache\Cache` interface so that we can build data collectors on top and show these stats in the web profiler.
ping @javiereguiluz

### CODE REVIEW

File: src/Symfony/Component/Cache/StatsInterface.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    /**
+     * Retrieves cache statistics from the data store.
+     *
+     * The statistics array has the following values:
+     * - hits: Number of keys that have been requested and found present.
+     * - misses: Number of items that have been requested and not found.
+     * - uptime: Time that the server is running.
+     * - memory_usage: Memory used by the server to store items.
+     * - memory_available: Memory allowed to use for storage.
```
Comment (javiereguiluz) [inline src/Symfony/Component/Cache/StatsInterface.php]: The description for `memory_available` looks confusing to me. Is it the total memory allowed or the remaining memory available?

File: src/Symfony/Component/Cache/StatsInterface.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    const STATS_MEMORY_USAGE = 'memory_usage';
+    const STATS_MEMORY_AVAILABLE = 'memory_available';
+
+    /**
+     * Retrieves cache statistics from the data store.
+     *
+     * The statistics array has the following values:
+     * - hits: Number of keys that have been requested and found present.
+     * - misses: Number of items that have been requested and not found.
+     * - uptime: Time that the server is running.
```
Comment (javiereguiluz) [inline src/Symfony/Component/Cache/StatsInterface.php]: I'd change this description by: `how long the server has been running (in seconds).`

File: src/Symfony/Component/Cache/StatsInterface.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    const STATS_UPTIME = 'uptime';
+    const STATS_MEMORY_USAGE = 'memory_usage';
+    const STATS_MEMORY_AVAILABLE = 'memory_available';
+
+    /**
+     * Retrieves cache statistics from the data store.
+     *
+     * The statistics array has the following values:
+     * - hits: Number of keys that have been requested and found present.
+     * - misses: Number of items that have been requested and not found.
```
Comment (javiereguiluz) [inline src/Symfony/Component/Cache/StatsInterface.php]: `hits: number of keys ...` but `misses: number of items ...`  Is this "keys" vs "items" difference intended?

File: src/Symfony/Component/Cache/StatsInterface.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+namespace Symfony\Component\Cache;
+
+/**
+ * Subset of the Doctrine\Common\Cache\Cache interface.
+ */
+interface StatsInterface
+{
+    const STATS_HITS = 'hits';
```
Comment (javiereguiluz) [inline src/Symfony/Component/Cache/StatsInterface.php]: I know I'm a bit obsessed with constant names ... but these names seem a bit redundant to me:

```php
StatsInterface::STATS_HITS
StatsInterface::STATS_UPTIME
StatsInterface::STATS_MEMORY_AVAILABLE
// ...
```

What about this instead?

```php
StatsInterface::HITS
StatsInterface::UPTIME
StatsInterface::MEMORY_AVAILABLE
// ...
```

**Comment by nicolas-grekas** [comment]
The comments in the interface are from the doctrine cache interface and I have exactly the same questions about them. When looking at the implementations there, it seems "uptime" is implemented as "start time", but is it true for all drivers? Some of them fetch stats info that are called "uptime" natively. Does this mean doctrine's stats return values change semantic from implem to implem?
@Ocramius & @Tobion, I'd really like to have your opinion on this and on the above @javiereguiluz's comments.

**Comment by fabpot** [comment]
I would instead create a wrapper (possibly in the existing proxy) instead of changing all implementation.

Overall, I'm not even sure it's something we need now. So, I'm -0 on this one.

**Comment by nicolas-grekas** [comment]
Updated
Status: needs review

**Comment by fabpot** [comment]
:+1:

**Comment by javiereguiluz** [comment]
:+1: 

This is definitely going to help us a lot to display this info in the new Cache profiler panel. Thanks @nicolas-grekas!

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**